src/os/ui/data/layercheckbox.js

```diff
@@ -0,0 +1,119 @@
+goog.module('os.ui.data.LayerCheckboxUI');
+
+const ConfirmUI = goog.require('os.ui.window.ConfirmUI');
+const Module = goog.require('os.ui.Module');
+const TriState = goog.require('os.structs.TriState');
+const TriStateCheckboxCtrl = goog.require('os.ui.TriStateCheckboxCtrl');
+const triStateCheckboxDirective = goog.require('os.ui.triStateCheckboxDirective');
+
+const TriStateTreeNode = goog.requireType('os.structs.TriStateTreeNode');
+
+
+/**
+ * Tristate checkbox extension for enabling layers.
+ *
+ * @return {angular.Directive}
+ */
+const directive = () => {
+  const dir = triStateCheckboxDirective();
+  dir.controller = Controller;
+  return dir;
+};
+
+
+/**
+ * The element tag for the directive.
+ * @type {string}
+ */
+const directiveTag = 'layercheckbox';
+
+
+/**
+ * Add the directive to the module
+ */
+Module.directive(directiveTag, [directive]);
+
+
+
+/**
+ * Controller for the layer checkbox.
+ * @unrestricted
+ */
+class Controller extends TriStateCheckboxCtrl {
+  /**
+   * Constructor.
+   * @param {!angular.Scope} $scope The Angular scope.
+   * @param {!angular.JQLite} $element The root DOM element.
+   * @ngInject
+   */
+  constructor($scope, $element) {
+    super($scope, $element);
+
+    /**
+     * The minimum count of layers before prompting the user.
+     * @type {number}
+     */
+    this.minCount = $scope['minCount'] || 10;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  toggle(e) {
+    if (this.scope) {
+      const item = /** @type {TriStateTreeNode} */ (this.scope['item']);
+
+      if (!this.getDisabled(item)) {
+        const children = item.getChildren();
+        const state = item.getState();
+
+        if (state == TriState.OFF) {
+          if (children && children.length > this.minCount) {
```

Comment:
There may be some specific situations we want to consider here:

- If this node has 5 folders with 5 layers each the warning would not be triggered but 25 layers would be added.
- If this node has 5 folders with 10 layers each, we would get one warning displayed by each of the 5 folders.

We can solve the first with a recursive child count, but the second would have to prevent the warning from also being displayed by the children in their `toggle` calls. One option might be making the `confirm` function here globally disable the warning before calling `toggleInternal` and enable again after the call.

Reply:
Will fix the recursive count issue.

For folders that might exceed the limit, I don't think the checkbox UI code is what gets called to activate them when the state change bubbles down the tree, it calls the `setState` function on the nodes. That's why I put this in the UI and not in the `setState` call: it should only ever be called by a click, never internal model changes. This also prevents it from being hit on application restore (which I ran into when I had it in the `setState` function initially).